### ISSUE TITLE
Re: MPU-401 intelligent mode #45 

### ISSUE THREAD

**Issue by jingber** [stsp/dosemu2#117]
I'm new to this project and read through the lengthy discussion regarding the MPU-401 "intelligent mode" discussion/debate.  https://github.com/stsp/dosemu2/issues/45

I feel this issue may warrant a re-evaluation:

While the bulk of games that install an MPU-401 interrupt handler may now "work", this is really not sufficient for full MPU-401 compatibility. There are features of the hardware such as programmable envelope generators which will never work without full emulation.  A good demonstration of these discrepancies is the opening sequence of Silpheed played through DOSEmu vs. DOSBox with the same underlying sequencer (MUNT through ALSA or an MT-32).  The DOSBox/MUNT combo is essentially indistinguishable from the actual Roland hardware to my ears where DOSEmu really needs some work.

I understand the goal of DOSEmu isn't to re-build DOSBox, but I feel this is an important feature as alot of older software needed this synth hardware, and a fork of DOSEmu already exists which provides this support (and needs some additional work).  This is really unique hardware that is not easily emulated, so any additional effort to preserve this capability would be greatly appreciated.

Please consider merging the MPU-401 fork and fixing any remaining issues to MPU-401 compatibility.

**Comment by jingber** [comment]
Just to add, the MPU-401 is really an "ugly" part of the IBM PC legacy, and it stinks that a "generic" DOS emulator needs to deal with this, but it's something we're stuck with now, for better or worse.  This standard is really intertwined with DOS gaming, and as such I feel it should be supported.

**Comment by runderwo** [comment]
The current state of the MPU401 code is
1. It mostly works, but has minor timing bugs, which might be fixed if I updated it to use some new functionality Stas added in the meantime.
2. It needs to be completely refactored anyway as the changes to dspio.c were/are considered unacceptable to merge.
3. There are a total of (now two) dosemu users who apparently care about this feature. :-)  Though, it seems that a lot of dosbox users care, judging from the forum posts...

**Comment by stsp** [comment]
> and a fork of DOSEmu already exists which provides this support

It still doesn't use munt.
So the best luck can be achieved using
munt as an alsa sequencer and making
dosemu to use alsa for midi, but this is
quite rough and difficult to configure.
So the fork you mention, doesn't fully solve
the problem, unless it starts using munt, I
suppose.

**Comment by runderwo** [comment]
Munt is not really difficult to configure; start the Qt console and execute one line of shell before starting dosemu. :-)
Munt could be used as a plugin too, but then configuring the ROM files is still a remaining manual step for the user.

**Comment by stsp** [comment]
Would be good as a first step, but at the
end we should be able to select gm/mt32
synth via blaster.com at run-time. Which
will likely only work in case of a plugin.

**Comment by runderwo** [comment]
Well, one could direct raw midi output to the alsa virmidi device node, instead of to the dosemu-midi fifo which requires launching a process to redirect it.  So that's an intermediate option.

**Comment by jingber** [comment]
I apologise if I didn't articulate this issue well --

This isn't a MUNT issue. 

This is a problem regarding any MT-32-style synth properly plumbed through ALSA that expects an MPU-401 host (either MUNT, or a bona-fide CM-32L/LAPC-I/MT-32, etc. device attached by way of a USB to MIDI adapter).

**Comment by jingber** [comment]
On a Debian-type distro, this is how to deal with MUNT --

1 - Install MUNT for source.  There will be an error regarding MT-32 ROMS.  Ignore it.

2-  In /etc/default, place the following as "mt32emu" (ie. /etc/default/mt32emu).

[CODE]# Default settings for mt32d. This file is sourced by /etc/systemd/system from
# /etc/systemd/system/mt32d*

MT32D_OPTS=

MT32D_DEVICE_INPUT='20:0'
MT32D_DEVICE_OUTPUT='128:0'[CODE]

In /etc/system/systemd, place the following two files:

/etc/system/systemd/mt32d.service

[CODE][Unit]
Description=mt32d MIDI emulator
Wants=dev-virmidi1.device
After=dev-virmidi1.device
Requires=mt32d-alsa.service

[Service]
EnvironmentFile=-/etc/default/mt32emu
ExecStart=/usr/local/bin/mt32d \
	$MT32D_OPTS
ExecReload=/bin/kill -HUP $MAINPID
KillMode=process
Restart=on-failure

[Install]
WantedBy=default.target[/CODE]

/etc/system/systemd/mt32d-alsa.service

[CODE][Unit]
Description=mt32d -> ALSA connector
After=mt32d.service
Requires=mt32d.service

[Service]
Type=oneshot
EnvironmentFile=-/etc/default/mt32emu
ExecStart=/bin/sleep 1
ExecStart=/usr/bin/aconnect \
	$MT32D_DEVICE_INPUT \
	$MT32D_DEVICE_OUTPUT
ExecStop=/usr/bin/aconnect --removeall
RemainAfterExit=yes

[Install] 
WantedBy=default.target[/CODE]

In /usr/share/mt32-rom-data place the MT-32/CM-32L control ROMS in this directory, and ensure all the file names are capitalized.

Reboot, and you're done.

**Comment by jingber** [comment]
It still doesn't use munt.

-- 

I'm not sure I'm following  ---- DOSemu can use ALSA for MIDI, and ALSA can be plumbed into MUNT, or a USB MIDI adapter connected to an MT-32, or a SC-55, or a Proteus, or a sewing machine, or anything else controlled by MIDI.

The problem is, the MPU-401 had some unique features, like the aforementioned  programmable envelope generators, that aren't present on Sound Blaster-type HW.  Certain DOS games use these features, and they either don't sound correct (in the case of Silpheed), or don't run at all (in the case of a handful of the DOS games listed in the bug # linked in the first post).

**Comment by jingber** [comment]
Best way to describe this --

1.  Get MUNT running in your distro, then get DOSBox and DOSEmu, both running through ALSA/MUNT.

2.  Play the into in Sylpheed in DOSBox and DOSEmu, and notice the difference in the sound through the emulated synth.

This is what is *should* should sound like:

https://www.youtube.com/watch?v=3n4nOiURgzg


It doesn't sound like this in DOSEmu, even tough the underlying synth is the same.  That's because DOSEmu lacks the features of a fully-fledged MPU-401.

T

**Comment by stsp** [comment]
Since dosemu have no way of knowing
what synth is attached to alsa (GM or MT32),
this will not work for a run-time selection.
If you install for example waveblaster2 daughterboard,
you can select between GM, GS and MT32
(with some sysexes I suppose).
So it may be good as a first step, but not more
than that.

**Comment by stsp** [comment]
> Well, one could direct raw midi output to the alsa virmidi device node, instead
> of to the dosemu-midi fifo which requires launching a process to redirect it.

Nobody uses this fifo these days.
Not sure why are you referring to it.

**Comment by stsp** [comment]
Another drawback of this approach is
that if you set a $_wav_file for recording,
midi will not be recorded (currently if you
use $_wav_file, GM is recorded).
But of course first lets have at least that,
then see.

**Comment by runderwo** [comment]
If you weren't talking about the fifo, then I'm not sure about the difficulty of configuration you're referring to.  One would simply set "virtual" instead of "default" for the ALSA rawmidi device, and whatever MIDI plumbing the user has created (real devices and/or emulator daemons) is automatically respected.  No, that is currently not exposed as a configuration option, but it should be trivial to do so.

In this case, it would be easy to automatically connect both Munt and Timidity daemons to the virmidi device, and then use the respective volume controls to determine which daemon output is actually audible.  This is similar in behavior to a chain of real MIDI devices.

**Comment by stsp** [comment]
@runderwo I am talking about the run-time
configuration, not dosemu.conf. Something
like WaveBlaster2 does. I don't see how do
you propose to achieve that.

**Comment by runderwo** [comment]
I don't understand what you're objecting to.  Maybe you have a different idea than me how you would use this.  If I have timidity and munt daemons running, the "runtime configuration" is the volume controls (through pulseaudio mixer), just like turning knobs on the real MIDI devices.  The devices all receive the MIDI commands, even if the instrument selection and sysexes are designed for a specific device.  If I am using a program that requires MT32, I mute timidity and unmute Munt.  If I'm using a GM program, I mute Munt and unmute Timidity.  If I don't want to waste cycles emulating a muted device, I can remove the ALSA virmidi connection for the device I don't care about so that the MIDI commands are not sent to it.

In any case, whether Munt is included as a plugin or not is a totally separate issue from whether the stateful MPU-401 interface is emulated.  It only happens that _most_ applications using MPU-401 hardware are 1) games that 2) are programmed for a MT-32.

I know you don't like this idea, and I'm open to using SoftMPU instead and dropping this patch as you suggested last year, since it's based on the same code from Dosbox anyway, but it would require implementing some vendor specific EMM extensions to trap the port I/O (and probably other things too).
https://github.com/bjt42/softmpu/blob/master/SRC/RESIDENT.ASM

**Comment by stsp** [comment]
I mean the selection from inside dosemu (eg blaster.com).
This is how WaveBlaster2 works - it has wb2mode.com
to select the needed emulation.
Also I don't think all devices receive the midi data.
Every midi synth module creates a new midi port.
And from munt readme:
https://github.com/munt/munt/blob/master/mt32emu_alsadrv/README.txt

When trying to play MT-32  midi streams use ALSA port 128:0.
128:1 attempts to emulate general midi support using the MT-32 mode

So even munt itself creates 2 midi ports.
Please provide the evidence that all devices
receive the midi stream in parallel, and you
can simply mute/unmute them. I've never
heard it works like that.

**Comment by runderwo** [comment]
WaveBlaster is a bad example because the MT-32 "emulation" simply selects patches that mimic MT-32's default patches.  The problem is, no program written for MT-32 (that is, pre-GM) uses the default patches. :-)  An actual MT-32 (or ROM-based emulation) is required in practice.

Ok, I understand your confusion now.  The part you're missing is virmidi.
`
$ aconnect -l
client 0: 'System' [type=kernel]
    0 'Timer           '
    1 'Announce        '
client 14: 'Midi Through' [type=kernel]
    0 'Midi Through Port-0'
client 20: 'Virtual Raw MIDI 1-0' [type=kernel]
    0 'VirMIDI 1-0     '
        Connecting To: 129:0, 128:0
client 21: 'Virtual Raw MIDI 1-1' [type=kernel]
    0 'VirMIDI 1-1     '
client 22: 'Virtual Raw MIDI 1-2' [type=kernel]
    0 'VirMIDI 1-2     '
client 23: 'Virtual Raw MIDI 1-3' [type=kernel]
    0 'VirMIDI 1-3     '
client 128: 'TiMidity' [type=user]
    0 'TiMidity port 0 '
        Connected From: 20:0
    1 'TiMidity port 1 '
    2 'TiMidity port 2 '
    3 'TiMidity port 3 '
client 129: 'Munt MT-32' [type=user]
    0 'Standard        '
        Connected From: 20:0
`

By making those connections (`aconnect 20:0 128:0; aconnect 20:0 129:0`), a MIDI bus is simulated for any data output to the "virtual" (as opposed to "default") ALSA rawmidi device.

Personally I like this setup because it can mix easily with real MIDI devices, and it's straightforward to create: 1) configure timidity as server, 2) mt32emu-qt + roms, 3) aconnect command, 4) set dosemu ALSA rawmidi device (currently hardcoded to "default" but can be made configurable).

But the plugin approach is arguably better for most users because the only setup is to find the MT32 ROM images, which cannot be distributed.  I think there is room for both approaches.

**Comment by stsp** [comment]
So this still doesn't work for selection within
dosemu. Also I don't think this is acceptable
at all: you want multiple midi renderers to run
and eat CPU, with just one being unmuted?
Also people may forget to mute the needed
thing etc. This is absolutely not user-friendly.
If you don't want the in-dosemu selection, you
can perhaps just move your mpu401 device
to the different IO port range. Then it can use
the different synth than the one of SB16.

**Comment by jingber** [comment]
I'd be tickled pink if DOSEmu would support the appropriate memory manager I/O hooks in EMM.SYS to run that MPU-401 emulator.

**Comment by jingber** [comment]
That said, I think the DOSBox MPU-401 fork should be pulled in and refined.

**Comment by runderwo** [comment]
Stas, even in the current iteration:
1. Multiprocess CPU utilization is less an issue with modern multicore systems
2. One can manually disconnect instead to muting if CPU utilization is an issue
Again, you're confusing "my fork currently doesn't do this" with "I don't want/think this is a bad idea".
The ALSA plugin could be used to list the available connections and connect to only the desired one via blaster.com.  It can not do that at present only because the code is not written.  MT-32 can be integrated as a plugin, but the code is not written, this requires external ROM files, and does not remove the need for the MPU-401 emulation currently under discussion.

The wrong idea that I thought you had is emulating MT-32 via any means other than the actual L/A synth emulator connected to an interrupt-driven MPU-401 emulator. (i.e. WaveBlaster style patch manipulation with UART hack)

Also, it's not feasible to move interrupt-driven MPU-401 to anything other than I/O 330 and IRQ 2 for compatibility reasons.

**Comment by runderwo** [comment]
@jingber I wouldn't merge it in its current state.  There is a known problem with timing and there are quite a few unrelated changes tagging along.  I welcome you to test it though and let me know how it works for you with different programs.
https://github.com/runderwo/dosemu2/tree/mpu401

**Comment by stsp** [comment]
> 1. Multiprocess CPU utilization is less an issue with modern multicore systems

No, its a big deal on a multi-user systems.

> The ALSA plugin could be used to list the available connections and connect
> to only the desired one via blaster.com.

Yes, this one is closer.
It might be better to connect to both and register
2 midi output plugins. blaster.com will then select
the needed one. We may need a flag in midi plugins
that will tell the midi core what kind of synth this
plugin provides.

WaveBlaster is only an example of simple selection.
We can even use its wb2mode.com sysexes
to do the switching, but that's all I was suggesting
to borrow from WaveBlaster's design.

> Also, it's not feasible to move interrupt-driven MPU-401
> to anything other than I/O 330 and IRQ 2 for compatibility reasons.

Maybe for testing purposes you can just route the
uart mode to GM synth and the intelligent mode to
MT32 synth. I would be surprised if intelligent mode
ever needs the GM synth, but uart mode needing
MT32 looks realistic.

**Comment by runderwo** [comment]
Well, multiuser systems are hardly the main use case for someone playing games with emulated MIDI music on dosemu.  In any case, I described my current setup only, and how one could use it without any further development other than the MPU-401 patch:  muting via pulseaudio mixer is easier than switching to a terminal to run aconnect whenever switching DOS programs with different MIDI requirements.

I'm not sure, what use would it be to emulate WB2 sysexes, except to use wb2mode.com specifically?  Does any other program switch WB2 automatically?

From what I can tell, it is a very rare game that 1. requires intelligent mode, 2. is not composed specifically for MT-32.  I can't think of an example, but someone probably could.

Basically, GM existed starting in 1991 and by that time, SB-MIDI, precursor to any MPU-401 UART emulation in a PC soundcard, was far more common than actual MPU-401 cards, which began in 1985 and were used with the MT-32 or FB-01.  Historically, it was obviously a mistake for any game to use intelligent mode, since any game could play soundtracks with the PIT, and no other MPU features are necessary for games.

On the other hand, some MT-32 games can use SB-MIDI (example: patched Sierra SCI games) and/or UART mode (example: unpatched Sierra SCI games), and many games programmed for GM via SB16's MIDI UART in the early 1990s also "support" MT-32, by remapping its default patches to approximate GM (example: later Sierra games).

**Comment by andrewbird** [comment]
Regarding the performance issues with this, it's not a problem for me providing it's not on by default with a simple sound setup. IMHO if you turn it on, and the performance sucks, you bear the  consequences.

**Comment by jingber** [comment]
OK, I made some videos that are really good examples of what is happening:

<a href="https://www.youtube.com/watch?v=1lnVUVUuIf8&feature=youtu.be">Silpheed into, bare metal</a>
<a href="https://www.youtube.com/watch?v=cDciXBDwb5c&feature=youtu.be">Silpheed intro, DOSBox with MUNT</a>
<a href="https://www.youtube.com/watch?v=g0vWqPlzC1Q&feature=youtu.be">Silpheed into, DOSEmu with MUNT <<-- FAIL </a>

**Comment by stsp** [comment]
@andrewbird Well, people DO care about the
CPU power. Think of notebooks. There were
already the reports about this:
https://github.com/stsp/dosemu2/issues/22
quote:
Playing it with cpu set to powersave on an old amd 2-core.
Using 'emumouse x 4 y 4' (not 8) makes the mouse move nice even with dosemu -S switch.
This way I do no longer have to delete or rename libao.so.
With dosBOX, when you are in a game with large map and many interrupted enemy paths (often the case in end-game state) settlers2 is stuttering, makes a pause for 2 seconds every 3 seconds or so.
Under dosemu it runs better and better.
/quote

Or here:
https://www.reddit.com/r/linux_gaming/comments/33y6x3/dosemu2_the_reincarnation_the_famous_dosemu/
quote:
Also consider that if it really is 10x faster, that's a lot more efficient than DOSBox, and on a phone or laptop, that could mean less battery usage.
/quote

So please, don't advocate for bad design decisions.
I am not saying we need to optimize every bit shift,
but running the whole midi rendering to just keep it
muted - no, users will not eat such things.

**Comment by stsp** [comment]
@jingber I am getting error on the first video
(2 others are accessible)

**Comment by stsp** [comment]
@jingber Are you sure the rendering for dosemu2
was done with munt on your video? It sounds as it
the default GM rendering was used. I am not sure
if @runderwo have implemented the needed config
knobs in his branch, so likely you tested the wrong
thing.

**Comment by andrewbird** [comment]
@stas, Perhaps I didn't make my point very well. I do care about performance very much so and what I wanted to say was that *IF* somebody were to implement some sound feature that did suck the last drop of power out of the machine, then I would very much like that to be a non default option that didn't impact normal performance for people who use dosemu for other tasks.
@jingber, I'm agnostic about this MIDI feature's inclusion, I don't really use sound so I'm no judge, though I did enjoy the videos. 

-- 
Andrew Bird <ajb@spheresystems.co.uk>

**Comment by stsp** [comment]
> I'm not sure, what use would it be to emulate WB2 sysexes, except to use wb2mode.com
> specifically? Does any other program switch WB2 automatically?

There are several games where you can select
WaveBlaster2 explicitly for music. They probably do.
But I am not sure. The only reason I am advocating
it, is actually because of my own setup: I had wb2
board 20 years ago, and I still have this user menu
in DosNavigator that allowed me in the past to
select the needed synth with wb2mode.com. The
compatibility with wb2mode is therefore of course
something entirely optional, but it is very simple also,
so why not?
Of course under dosemu you have plenty of other
choices, like the use of the doshelper interrupt instead,
or whatever.

> Basically, GM existed starting in 1991 and by that time, SB-MIDI, precursor
> to any MPU-401 UART emulation in a PC soundcard, was far more common
> than actual MPU-401 cards

If I parse that correctly, you are saying that
sb-midi was more common that mpu-401.
Since sb-midi was supported only on SB cards
and no software ever used it, I wonder how is
that possible. Well, maybe you mean its implementation
was common as an SB cards were common, but
if we are talking about its use - it was not common at all.

In any case, perhaps someone (me?) have to at
least implement a config selection for the midi driver,
so that @jingber can at least test the right thing.

**Comment by stsp** [comment]
@andrewbird I meant this sentence of your:

IMHO if you turn it on, and the performance sucks, you bear the consequences.
---

I disagree with this, because in this particular case
the consequences are highly unexpected: you only
wanted to enable 1 rendering engine, but got CPU
power drop for 2! We can't surprise the users as that.
I still remember how I was tuning the timidity parameters
on my old athlon700 10 years ago, so that dosemu
can at least barely run while timidity is synthesing.
The CPUs are now much more powerful, but still the
midi renderer will mostly eat one of your CPU cores.
Allowing this to happen in vain - is not the right decision.

**Comment by andrewbird** [comment]
Yep not enough thought went into constructing that sentence! It just flew out coming from the viewpoint of I'm very unlikely to ever turn it on, so I wouldn't really care one way or the other.

I'd better lie down now - not intending to stir things things up!

On Fri, 11 Dec 2015 04:58:22 -0800
Stas Sergeev <notifications@github.com> wrote:

> @andrewbird I meant this sentence of your:
> 
> IMHO if you turn it on, and the performance sucks, you bear the consequences.
> ---
> 
> I disagree with this, because in this particular case
> the consequences are highly unexpected: you only
> wanted to enable 1 rendering engine, but got CPU
> power drop for 2! We can't surprise the users as that.
> I still remember how I was tuning the timidity parameters
> on my old athlon700 10 years ago, so that dosemu
> can at least barely run while timidity is synthesing.
> The CPUs are now much more powerful, but still the
> midi renderer will mostly eat one of your CPU cores.
> Allowing this to happen in vain - is not the right decision.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/stsp/dosemu2/issues/117#issuecomment-163931163


-- 
Andrew Bird <ajb@spheresystems.co.uk>

**Comment by stsp** [comment]
I added the $_midi_driver config, so that the
alsa driver can be selected explicitly.
Ryan, does your tree need this? According to
the video, it seems so. Then please pull.

**Comment by runderwo** [comment]
Stas, it's the user (me) that would aconnect to both instances.  This is the setup that is convenient for me, a regular MIDI user who also connects real MIDI controllers and instruments, and who uses pulseaudio to control the emulator sound levels.  A user that does not want to waste the CPU can run one or both daemons and only connect to the one desired.  That is a manual process right now.  The _easiest_ thing to do is connect to both right now and set the volume levels.  If someone wants to write a Munt plugin to remove the daemon and to choose the MIDI instrument through a file or blaster.com instead, they can do so, but it does not remove the need for simulating the MPU-401.  I don't know why this particular debate continues.  I understand you are not a fan of my configuration but it is a simple fact that this is what I am using.

Also, I don't understand why a program, if WB2 were selected in the configuration, would want to change its mode from GM to something else.  Why would it not just set GM and use it as a GM instrument?  Maybe if the soundtrack were specifically composed for GS?  I don't know of such a program, maybe you do?

SB-MIDI was used by every SCI Sierra game with a MT-32 soundtrack and the patch.  This was an easy change for them to make because the sequencer is in the game driver (they didn't actually use sequencer features of MPU-401).  I don't know of any game that used intelligent mode after this patch became available, so I see this as the big shift away from MPU-401 (Sierra was previously selling these devices paired with MT32)

The MIDI output driver you added is not bad, but isn't related in this case.  We need to select the ALSA rawmidi  output driver to be able to ask for a "virtual" port instead of "default" (hw), and to select the port to which it will connect (i.e. connect dosemu's 130:0 to timidity 128:0, or munt 129:0).  The latter should be runtime configuration.

@jingber That sounds like timidity output, not Munt.  Could you post your dosemu.conf sound section?

**Comment by runderwo** [comment]
@stsp I rebased my fork, if it helps somehow

**Comment by stsp** [comment]
> A user that does not want to waste the CPU can run one or both
> daemons and only connect to the one desired.

... without an ability to select it with blaster.com/wb2mode.com.

> If someone wants to write a Munt plugin to remove the daemon
> and to choose the MIDI instrument through a file or blaster.com instead

You can simply register 2 alsa midi plugins, one
for GM and one for MT32. Or you can select between
fluidsynth for GM and alsa for MUNT. I don't see why
you need the plugin code for this. (but you will definitely
need a plugin once you want $_wav_file to capture the synth)

> Also, I don't understand why a program, if WB2 were selected in
> the configuration, would want to change its mode from GM to something else.

Maybe it wants to change to GM from something else? ;)

> The MIDI output driver you added is not bad, but isn't related in this case.
> We need to select the ALSA rawmidi output driver

Great so how do you select it without my patch?
My patch does exactly that.

> to be able to ask for a "virtual" port instead of "default" (hw), and to
> select the port to which it will connect (i.e. connect dosemu's 130:0
> to timidity 128:0, or munt 129:0). The latter should be runtime configuration.

AFAIK you can define "default" in /etc/asound.conf or .asoundrc in any way you want.
At least that works with pcm. Are you sure for midi its impossible
to redirect the "default"?

**Comment by runderwo** [comment]
Yes, without the ability *at present* to select with blaster.com.  Again, this patch is only about the MPU401 functionality which, again, has nothing to do with how the MIDI instruments and/or emulators are plumbed, which is a completely separate question.  Someone who is interested in testing can mimic my setup, or he can refuse to test.

I still don't understand the use case for supporting WB2 sysex.  Do you mean that Timidity would be in GS state and we would support the sysex so a WB2-supporting game would automatically remap it back to GM?  Please, tell me how you would use this.

Regarding rawmidi output driver:
http://www.alsa-project.org/alsa-doc/alsa-lib/rawmidi.html#rawmidi_virt
e.g.
snd_rawmidi_open(&read_handle, &write_handle, "virtual", 0);
You can see the patch in my tree, I separated it.
https://github.com/runderwo/dosemu2/commit/83357c276bf08f9b3b04d63c69d96b76162ab8fe

I have no idea about reconfiguring asoundrc to redirect the default to virtual.

**Comment by stsp** [comment]
> state and we would support the sysex so a WB2-supporting
> game would automatically remap it back to GM?

No, I was only answering your question why some
_game_ may want to use it. I am not saying it should,
neither am I saying we should set wb2 in any game's
setup. But we could still use wb2mode.com since its
very simple. And yes, the game may issue that sysex
if it thinks wb2 is in mt32 mode, to switch back to GM.

> I have no idea about reconfiguring asoundrc to redirect the default to virtual.

But that needs to be investigated.
I reconfigured "default" for pcm many times, and
I expect it works the same for midi, but someone
needs to try and see.

**Comment by stsp** [comment]
> @jingber That sounds like timidity output, not Munt. Could you post your dosemu.conf
> sound section?

Ryan, what option do you expect to see there,
if the only relevant option was just implemented
by me?

**Comment by runderwo** [comment]
If it's anything other than alsa, it can't possibly work at present, since Munt has to be connected through virmidi.

**Comment by stsp** [comment]
Ryan, there is simply no such option.
Maybe you compile-time disabled all but alsa,
but it doesn't mean others did too.

**Comment by runderwo** [comment]
The ability to use virmidi is in my branch only.

**Comment by jingber** [comment]
The DOSEmu/MUNT video is infact using MUNT through ALSA.  I don't have timidity++ on this system.

It's easy enough to verify anyhow - stopping mt32d will silence the output.

**Comment by jingber** [comment]
The 9th post in is the configuration I use for the MUNT/ALSA combo.

**Comment by runderwo** [comment]
I don't know about that config.  I'm using munt directly from http://github.com/munt.  I start mt32emu-qt:
client 129: 'Munt MT-32' [type=user]
    0 'Standard        '

Then I start dosemu and aconnect it before proceeding.  Silpheed works except for a small part of the intro is missing.  Are you sure you're using the mpu401 branch from my github?

**Comment by stsp** [comment]
> The ability to use virmidi is in my branch only.

Ryan, how do you force alsa if there is no such option?
And I don't even see any black magic in your branch to
force it. Have you compiled out all other midi plugins or what?
Please add the -D+S log.

**Comment by stsp** [comment]
> The DOSEmu/MUNT video is infact using MUNT through ALSA.
> I don't have timidity++ on this system.

Do you have fluidsynth there?
dosemu doesn't pick up alsa if there are any other renderers,
so I am very surprised it does so in this thread.

**Comment by runderwo** [comment]
I'm forcing a component of alsa, not alsa itself, see the patch I linked.
ALSA itself is used by default, at least with using compiletimesettings.devel on my system.

[00000000] @ FFFF:0000 SB: SB Initialisation
[00000000] @ FFFF:0000 PORT: registered "SB Emulation" handle 0x1c [0x0220-0x0233] fd=-1
[00000000] @ FFFF:0000 SB: Mixer init
[00000000] @ FFFF:0000 SB: OPL3 Initialization
[00000000] @ FFFF:0000 PORT: registered "OPL3" handle 0x1d [0x0388-0x038b] fd=-1
[00000000] @ FFFF:0000 PCM: Stream 0 allocated for "Adlib"
[00000000] @ FFFF:0000 MPU401: MPU-401 Initialisation
[00000000] @ FFFF:0000 PORT: registered "Midi Emulation" handle 0x1e [0x0330-0x0331] fd=-1
[00000000] MPU401: MPU-401 Initialisation - Base 0x330, IRQ 2@ FFFF:0000 
[00000000] @ FFFF:0000 SB: Initialisation - Base 0x220
[00000000] @ FFFF:0000 PCM: registering player: SB REC
[00000000] @ FFFF:0000 PCM: init
[00000000] @ FFFF:0000 PCM: registering player: Sound Output: libao
[00000000] @ FFFF:0000 PCM: registering player: Sound Output: libao wav writer
[00000000] @ FFFF:0000 PCM: registering recorder: Sounds input: alsa
[00000000] @ FFFF:0000 no PCM effect processors initialized
[00000000] @ FFFF:0000 PCM: Initializing selected plugin: Sound Output: libao: Failed
[00000000] @ FFFF:0000 PCM: Initializing pass-through plugin: SB REC: OK
[00000000] @ FFFF:0000 PCM: Stream 1 allocated for "PCM LINE IN"
[00000000] @ FFFF:0000 PCM: Initializing plugin: Sounds input: alsa (w=0): OK
[00000000] @ FFFF:0000 PCM: Stream 2 allocated for "SB DAC"
[00000000] @ FFFF:0000 PCM: setting flag 1 for stream 2 (SB DAC)
[00000000] @ FFFF:0000 PCM: Stream 3 allocated for "SB DMA"
[00000000] @ FFFF:0000 PCM: setting flag 4 for stream 3 (SB DMA)
[00000000] @ FFFF:0000 MPU401: dspio_mpu401_reset
[00000000] @ FFFF:0000 MIDI Output: named pipe: unable to open /home/nemesis/.dosemu/run/dosemu-midi for writing (No such device or address), will continue trying
[00000000] @ FFFF:0000 PCM: Initializing pass-through plugin: MIDI Output: named pipe: OK
[00000000] @ FFFF:0000 PCM: Initializing plugin: MIDI Output: TiMidity++ plugin (w=2): Failed
[00000000] @ FFFF:0000 PCM: Bypassing plugin: MIDI Output: OSS sequencer: (0 < 1)
[00000000] @ FFFF:0000 PCM: Initializing plugin: MIDI Output: ALSA device (w=1): OK
[00000000] @ FFFF:0000 MIDI Input: named pipe: unable to open /home/nemesis/.dosemu/run/dosemu-midi_in for reading: No such file or directory
[00000000] @ FFFF:0000 PCM: Initializing plugin: MIDI Input: named pipe (w=0): Failed

**Comment by jingber** [comment]
Guys, there are no other renderers in use.  This is absolutely MUNT trough ALSA.  I don't know what I can do besides recording a video and showing my configuration???


client 0: 'System' [type=kernel]
    0 'Timer           '
    1 'Announce        '
client 14: 'Midi Through' [type=kernel]
    0 'Midi Through Port-0'
client 20: 'Virtual Raw MIDI 1-0' [type=kernel]
    0 'VirMIDI 1-0     '
	Connecting To: 128:0
client 21: 'Virtual Raw MIDI 1-1' [type=kernel]
    0 'VirMIDI 1-1     '
client 22: 'Virtual Raw MIDI 1-2' [type=kernel]
    0 'VirMIDI 1-2     '
client 23: 'Virtual Raw MIDI 1-3' [type=kernel]
    0 'VirMIDI 1-3     '
client 128: 'MT-32' [type=user]
    0 'Standard        '
	Connected From: 20:0
    1 'GM Emulation    '

**Comment by jingber** [comment]
@runderwo - that configuration just starts the synth as a system-wide daemon, rather than from the Qt interface.

**Comment by jingber** [comment]
@runderwo - I am using DOSEmu from head.  I have not tried your patch.  I Will do so now and report back.

**Comment by stsp** [comment]
> ALSA itself is used by default, at least with using compiletimesettings.devel on my system.

Which is the most strange part.
Well I see timidity init fails for you for some
reason, and fluidsynth seems to be not compiled
in (no signs of it in the log). Well, timidity usually
fails because of an incompatible version, but what
was wrong with fluidsynth? fluidsynth also have disabled
the dosemu support in their git, but AFAIK no distro
is yet providing the dosemu-incompatible fluidsynth.
Or have this already changed? :( What version do
you have?

**Comment by runderwo** [comment]
Ok, here's what's wrong:

client 128: 'MT-32' [type=user]
0 'Standard '
Connected From: 20:0
1 'GM Emulation '

You've connected the virtual raw midi port (which ALSA will use by default with rawmidi when there's no MIDI hardware) to MT-32's GM emulation.  That won't work... aconnect -d that connection and then aconnect 20:0 128:0 and try again.

**Comment by runderwo** [comment]
@stsp 
ii  libfluidsynth1:amd64                           1.1.6-2build1                amd64                        Real-time MIDI software synthesizer (runtime library)

Although, I did not have the -dev package installed.  Now that I have it installed and rebuilt dosemu:
[00000000] @ FFFF:0000 MIDI Output: named pipe: unable to open /home/nemesis/.dosemu/run/dosemu-midi for writing (No such device or address), will continue trying
[00000000] @ FFFF:0000 PCM: Initializing pass-through plugin: MIDI Output: named pipe: OK
[00000000] @ FFFF:0000 PCM: Bypassing plugin: MIDI Output: TiMidity++ plugin: (2 < 3)
[00000000] @ FFFF:0000 Your fluidsynth is too old
[00000000] @ FFFF:0000 fluidsynth: loaded soundfont /usr/share/sounds/sf2/FluidR3_GM.sf2 ID=1
[00000000] @ FFFF:0000 PCM: Stream 4 allocated for "MIDI"
[00000000] @ FFFF:0000 PCM: Initializing plugin: MIDI Output: FluidSynth device (w=3): OK
[00000000] @ FFFF:0000 MIDI Input: named pipe: unable to open /home/nemesis/.dosemu/run/dosemu-midi_in for reading: No such file or directory
[00000000] @ FFFF:0000 PCM: Initializing plugin: MIDI Input: named pipe (w=0): Failed

**Comment by stsp** [comment]
So no alsa any more? Finally! :)

**Comment by runderwo** [comment]
Yes, I guess it was just luck that alsa was used by default since nothing else was available first.
Anyway, now $_midi_driver has the expected effect.

**Comment by stsp** [comment]
> @runderwo - I am using DOSEmu from head. I have not tried your patch.
> I Will do so now and report back.

Not so fast!
You had a wrong rendered - that was obvious
from your video, and it seems Ryan have already
identified the cause - wrong midi port on munt.
Please stick to the main branch but do what
Ryan said to fix the connection. We need to see
the result from that test.

**Comment by stsp** [comment]
> Yes, I guess it was just luck that alsa was used by default since nothing else was available first.
> Anyway, now $_midi_driver has the expected effect.

Maybe you can also find out why timidity fails for you?
Its quite bad for me to see already 2 people in this
thread for whom dosemu goes as down the stack as
to use alsa. We need to see if that can be improved -
alsa should never go in by default.

**Comment by stsp** [comment]
> [00000000] @ FFFF:0000 PCM: Initializing selected plugin: Sound Output: libao: Failed

Oh, and this is absolutely nasty too!
@runderwo, you don't even have the PCM audio, or do you?
Please open a bug report if that's the case.
We need to find out why libao fails for you.

**Comment by runderwo** [comment]
@stsp He's using a script to create the virmidi connections.  This is a lot of setup that is unnecessary if we support the ALSA virmidi output ourselves as in the one-line patch I showed you.

I'll check audio, I haven't done so in a while.  Debugging timidity is a bit low on the list but I'll take a look :/

**Comment by jingber** [comment]
@runderwo - You're not looking at that list correctly.  20 is connected to 128.

**Comment by runderwo** [comment]
@jingber you're right, I just checked mine again.  Sorry about that.  I don't know what's wrong then.  Your munt is quite strange, since mine (latest from github) doesn't even have a GM emulation option.

**Comment by jingber** [comment]
@runderwo - Let me trying buiding the Qt version so at least we're all the same sheet of music (yuck, yuck)...

**Comment by stsp** [comment]
> Debugging timidity is a bit low on the list but I'll take a look :/

Just post the version of it.
Most timidity versions are still incompatible
with dosemu (and fluidsynth will be incompatible
since its next version)

**Comment by runderwo** [comment]
It's 2.13.2-40.2build1 debian package (debian hasn't packaged 2.14.0 yet, only existed for 3 years) 8-)

**Comment by jingber** [comment]
@stsp - I don't agree with this assessment.  MUNT through ALSA is the only external MIDI renderer on this system and it's setup correctly.  

This is an issue with Silpheed, as it's doing some extra stuff with the MPU-401 that other games don't.

I have the real hardware right here for comparison, I know what it sounds like, and it works properly for most software.

**Comment by runderwo** [comment]
@jingber Something is wrong/different with your setup, Silpheed works fine for me, aside from a missing part in the intro.  Believe me, I listened to this soundtrack since I was 8 years old.  We just need to find the difference.

**Comment by runderwo** [comment]
@stsp DOS4GW games (tried wacky wheels and doom) are hanging randomly.  Could this be related to recent kvm changes?

**Comment by jingber** [comment]
I'm building the Qt version of MUNT now.  This way, we can see the front panel display so there will be no question the correct synth is in use.

My MUNT is from github maybe a month or so ago.  The only difference is, I'm using the daemonized version rather than the Qt version.

**Comment by runderwo** [comment]
@jingber can you md5sum the roms you are using?

**Comment by jingber** [comment]
@runderwo -

bfff32b6144c1d706109accb6e6b1113  CM32L_CONTROL.ROM
08cdcfa0ed93e9cb16afa76e6ac5f0a4  CM32L_PCM.ROM
9513fec4f09a7d327748340ce3a2a59b  MT32_CONTROL.ROM
89e42e386e82e0cacb4a2704a03706ca  MT32_PCM.ROM

**Comment by runderwo** [comment]
ver1.07 10 Oct, 87 is the version string in my MT32_CONTROL.ROM

**Comment by jingber** [comment]
Just to reiterate, it's only falling over with Silpheed.  I don't have problems with any other games - everything else that is able work under DOSEmu sounds perfect.  

DOSBox works through MUNT with these same ROMS and plays the music like champ.

**Comment by runderwo** [comment]
I don't know why there would be a difference with Silpheed.  Maybe some kind of race condition that doesn't happen on my system.  Thanks for clarifying that other things are working, that helps.

**Comment by jingber** [comment]
MUNT will use the CM32L ROMs if present over the MT-32 ROMS.  Do you have both sets in the directory, or just the MT-32 ROMs?

**Comment by runderwo** [comment]
Both are present, but the MT32 ones are checked in munt ROM selection (Qt app)

**Comment by jingber** [comment]
Where is the version code in the ROM?  Grepping a 'strings' output on the file lists "8096 MIDI monitor on SI-61A    87/03/30", but that may the version code for that particular routine.  There's also a reference to "30 Sep. 88" which attributes authors which I think refers to the creators of the instrument patches.

**Comment by jingber** [comment]
Let me move the CM32 ROMs out of the way...

**Comment by runderwo** [comment]
@stsp PCM audio works in wolf3d, I guess through SDL.

**Comment by jingber** [comment]
@runderwo - We have a different control ROM, but the CM32L ROMs produce the same effect.

**Comment by runderwo** [comment]
@jingber I can't explain it.  Maybe you could look at the port I/O when it uploads instrument patches.  Maybe something is going wrong in that spot but it continues anyway and can somehow play music.  You can log with dosemu.bin -D+S -o log.txt

**Comment by stsp** [comment]
@jingber just to make sure: have you tried Ryan's
fork after all? Maybe the problem only happens on
the main branch?

**Comment by runderwo** [comment]
That's true, since with the UART hack it's possible Sierra games would work, since they don't care about any actual hardware features.

**Comment by stsp** [comment]
> @stsp PCM audio works in wolf3d, I guess through SDL.

Without any signs of SDL in your log and with the failed libao? :(

**Comment by runderwo** [comment]
No, I had to reinstall the SDL lib and rebuild.  After doing so, SDL audio works.  But libao still mysteriously fails.

**Comment by stsp** [comment]
@runderwo Please open a bug report then.

**Comment by runderwo** [comment]
Don't worry, I'll find out why when I'm done with this problem at hand...

**Comment by runderwo** [comment]
@stsp the bug is passing "buffer_time" as an option to libao is invalid for the pulseaudio plugin.

**Comment by jingber** [comment]
All,

I put a new video up, directly comparing DOSBox vs DOSEmu with the same MUNT Qt synth.  I pulled this MUNT directly from Github right before I made the video.


<a href="">Silpheed DOSBox vs DOSEmu through MUNT.</a>


If it's not playing for you, it's Google's cache latency - try it again in a few mins.

This should be **absolutely** convincing my setup is correct and DOSEmu is doing something different than DOSBox with regard to Silpheed, at least on *my* system.  It should also validate my sanity :)  

<a href="https://www.youtube.com/watch?v=SKpLuO4WunY">Silpheed DOSBox vs DOSEmu through MUNT.</a>

Shaky camera.  We run a very low-tech operation here.

**Comment by jingber** [comment]
@stsp -

I have, and all programs using an MPU-401 lock up immediately when the first MIDI commands are sent to the hardware.

I'm playing with this now, and I'll send the logs shortly if I can't figure this out.

**Comment by jingber** [comment]
@runderwo - 

We're running the same synth now, let's make sure we're running the same version of the game.  I'm running what is apparently some type of IBM promotional version that uses SB-type HW in place of the PC Speaker for game audio.

<a href="https://www.youtube.com/watch?v=9uO9XDe3CL4">Silpheed on dual sound card version under DOSBox/MUNT</a)


(You can tell if you're running the dual sound card version if the "talking head" is intelligible -- "I am Xacalyte, Emperor of the Universe..."), and the game audio is rendered on the SB.

Here's what I'm running -- (http://home.ix.netcom.com/~jjingber/silpheed.tar.bz2).  Make sure to run run the installer if this is something new.



0af199c08db67ec9f2c660bb1a667c4c  ADL.DRV
1904089e78dc593d51bd51eb8e61fcb9  AUDBLAST.DRV
b27147b87df3e9d8f0c2bf48b6359a1f  CARTOON.EGA
0cc6d6a78238ce847c92e669d1af8138  D1.EGA
e93a50029a78192a892115c7622c078e  DEFAULT.SAV
72561a9ffd26f75283f2bec6a88d628c  D.EGA
e46bdf6a7d21b01a05acb9862fc55e69  EFC001
d7911729bb837d7952a2ed5413b368e1  EFC002
f1cb71e754c7310f9b85dcf776ede3ea  EFC004
986f3fd068277f7632915c37f5380344  EFC005
0f3ab2d78d17fcb147e6048244a82dec  EFC009
6a13a175f2448f4a927f5804436b91e7  EFC010
14fad812a758c5e1bf06511e23d95df1  EFC011
0be8a76d676ba7388eb32804884b2499  EFC012
74fa122c40469524e82df2e017227a33  EFC015
7cae79adfc84c9d4a34e21f70413fd30  EFC017
61f35eec8998802abe2118b26e0c9cfa  EXISTS.COM
358131d9459ce17c40f869d64f8c3361  GDEMO.EGA
7c30f68a055332b3fe71bdf6505f5378  GRAPHICS.EGA
9ebde3a62e99a49b9edaab912222cc51  IBMCARD.DRV
a2a54f34b4b1cb17935ad535aab74c3e  ID
c9ae358d63df05d738e6819d606e625b  IMF.DRV
5ee712fbeac36ef27fadcf8d763425b1  INSTALL.EXE
4315bb37000cb56f385541021aad26ea  __INSTH.BAT
1ec125832887e0448272bdc48429462d  JOYSTICK.DRV
a0eb6e3f4851488d68f31913f48a0670  MT32.DRV
afbca36ded2e7c5fd065c7a8c78988e9  MT32IBM.DRV
fc5b4520f40d364f112edf3438fc8e22  PAT001
22bcd42b45f1a90da980eefda57f8a93  PAT002
05df3d5e3088959ebf9c1b1e58d34025  PAT003
f61e016783f873af60c6ff0b61211468  PAT004
7b1a50b8388f4176c80b84b2d647f74e  PAT101
a1c1920ca434032905c9c9e5f9fcbf9e  PCEFFECT.DRV
4367dfd887609ec533f29878740da8c2  README.BAT
3a0c37ee4e3a1f5d3fd64aa872ef8353  RES001
93d6caa0eafcd4b68626eef18806aeb6  RES002
a3da331f687ca103472c3b370a7d3411  RES003
b6b20f28c1e66a327ae75bb69ccbfdd2  RES004
e28d4fd2382da17d6e69c06bde512c45  RES005
da2fe5514def20ff548dd62db47a28e9  RES006
f20a15fc5838d70ebfdf1617ae90753e  RES007
f77c7c028ea7b3dfe83b91b689a26014  RES008
95418908b0b63cbc566ddfa772bb1c5d  RES009
8ae5c07257494d139de3ab001dd07899  RES010
da0a13e5c351c3419ea0b673233d51c8  RES011
67a9c192625b5409b61aeb42f32cccc9  RES020
0c0a09ebc78e87b6911bff84462a5b25  RES021
ae87e6ff42bd153b57020e81b5031545  RES022
af29c3fd7d35824bc65ef75c11cbf21a  RESOURCE.CFG
2b9c74743f99170fc3229c92c065922b  SAMPLE.DRV
eb4248958dfe7f70fa9316e23a665529  S.EXE
8a3a4bfe672c0033549c56945e479966  SIERRA.EXE
6e29899cb0eb9e64a802139c90480902  SILPHIGH.SAV
6e29899cb0eb9e64a802139c90480902  SILPLAST.SAV
9fc558d75da93f72e3469b4534706b37  STD.DRV

**Comment by jingber** [comment]
What happens if you actually try to run this game thorugh Timidity, etc:  <a href="https://www.youtube.com/watch?v=ydZZAk1bN6g">Silpheed on DOSBox/Timidity</a>

ED:  I sounded testier in that video than I meant to... :).  Time to pack it in.

**Comment by stsp** [comment]
@runderwo Thanks!
What libao version do you have?
Please note that I _need_ to set "buffer_time" because
of this bug:
https://bugzilla.redhat.com/show_bug.cgi?id=1193688
libao is exceptionally lame. Which is what I wrote in the
code comments in many places.
Please see here:
https://xiph.org/ao/doc/ao_open_live.html
It explicitly says:
"Unsupported options are ignored."
Which is not what we see. Could you please report the bug
to debian bts?
I can sure provide the work-around (and I'll do), but the
whole libao support is a mess of hacks and workarounds.
I am really considering to drop it, its just the horrible piece
of code.

**Comment by stsp** [comment]
@jingber So the only guess here would be that
some sysex is "eaten" by dosemu, that should
program the mt32 for the right thing. Please
make the -D9+S log.
Also I am still unsure, is the problem appear same
on main branch and on @runderwo's fork?

**Comment by stsp** [comment]
> I have, and all programs using an MPU-401 lock up immediately when the first
> MIDI commands are sent to the hardware (MPU-401 on IRQ 2/IO 330h, SB on IRQ 5/220h/DMA 1)

You mean, on main branch? And no such problem on
@runderwo's branch. right? Or?

**Comment by runderwo** [comment]
@stsp Finally, I worked around the timing problems.  Gateway plays with the proper tempo (on average) now and catches up if IRQ servicing lags.  That was not easy :/

@jingber I will look at this version of Silpheed now.

**Comment by runderwo** [comment]
@jingber interestingly, if I use the MT32.DRV from 1989 Silpheed with the PS/1 version of the game, the music works correctly...

**Comment by runderwo** [comment]
@jingber Found the difference; MT32.DRV from 1989 Silpheed uses intelligent mode (working correctly), while MT32.DRV from PS/1 Silpheed uses UART mode, which is not working correctly in this case.

**Comment by runderwo** [comment]
I tried adding some 1ms sleeps in between MIDI output bytes in case this was an overflow issue with the MT-32 ROM, but it didn't help.  So I'm not sure what the problem is right now.

**Comment by jingber** [comment]
Will do, ASAP.

**Comment by jingber** [comment]
@stsp -

All my tests have been with DOSEMU2 head.

I briefly tried to use runderwo's MPU-401 branch, but I wasn't successful at running anything using an MPU-401, but I'll mess around with this some more.

Keep in mind, the "PS/1 talking head" version does infact run under DOSBox, so, there's still something different going on in these emulators.

**Comment by stsp** [comment]
So we have a problem here even with UART mode?
Interesting, I wonder if I can reproduce it too then...

@runderwo please fill in the libao bugreport to debian,
as I'd like to reference it in the comments when adding
yet another silly work-around.

**Comment by jingber** [comment]
OK --

(Notice where the logfiles for the "working" runs are several orders of magnitude larger than the non-working runs).


The working MT32DRV came from from Space Quest III.


<a href="http://home.ix.netcom.com/~jjingber/MT32.DRV-Working">MT32.DRV  -- Working in DOSEmu and DOSBox</a>
<a href="http://home.ix.netcom.com/~jjingber/MT32.DRV-Working.log">MT32.DRV -- Working in DOSEmu and DOSBoxLog logfile </a>

<a href="http://home.ix.netcom.com/~jjingber/MT32.DRV-Broken">MT32.DRV  -- Not Working in DOSEMu, but working in DOSBox</a>
<a href="http://home.ix.netcom.com/~jjingber/MT32.DRV-Broken.log">MT32.DRV -- Not Working in DOSEmu, but working in DOSBox logilfe </a>

**Comment by jingber** [comment]
I timed both of these runs and terminated them with a script, so the file size difference is legit.

**Comment by stsp** [comment]
@jingber I believe both logs were interrupted before
the music started to play. Don't worry about the size,
please make a bigger logs.

**Comment by stsp** [comment]
although of course if the problem is now reproduceable,
someone else can create the log too.

**Comment by jingber** [comment]
@stsp - This is an old game, so I don't know if there's any other way to shut down the game short of closing the window.

**Comment by stsp** [comment]
Why you can't close the window much later?

**Comment by stsp** [comment]
> Why you can't close the window much later?

I know why, because then it crashes.
This is now fixed: you can now close dosemu window
even while the intro music is playing.

**Comment by stsp** [comment]
So the difference is that the driver that uses
the UART mode, initially passes lots of sysexes.
dosbox sends sysexes to alsa much differently
than dosemu does. dosbox uses snd_seq_ev_set_sysex()
and dosemu uses snd_rawmidi_write().
Ryan, could you somehow check on the munt
side if the sysexes are passed correctly?

**Comment by stsp** [comment]
https://code.google.com/p/midisnoop/

**Comment by stsp** [comment]
Could anyone please remove SND_RAWMIDI_NONBLOCK
in mid_o_alsa.c and re-check?

**Comment by stsp** [comment]
Please re-test with current devel.

**Comment by runderwo** [comment]
@stsp https://bugs.launchpad.net/ubuntu/+source/libao/+bug/1525776
@jingber Are you sure you used the 'mpu401' branch?  Please git branch -l or git status in your checked-out repo.

**Comment by stsp** [comment]
Ryan, I think @jingber said a few times here
that he is testing main branch, not your mpu401
branch. That's because the problematic driver
uses UART mode, and the working driver uses
d0 command of intelligent mode, which is similar.
So lets concentrate on fixing the problem in one
branch, this thread is already a bit chaotic.
I am pretty sure we have problems with passing
sysexes through alsa.

**Comment by stsp** [comment]
@runderwo Could you please re-test libao?
I applied the w/a.

**Comment by runderwo** [comment]
@stsp the UART/WSD sysex upload in PS/1 silpheed is still broken.  But libao now works!

**Comment by stsp** [comment]
> @stsp the UART sysex upload in PS/1 silpheed is still broken.

Can you somehow check on munt's side if the sysexes
are passed well? Either with some logging in munt or with
midisnoop perhaps? I haven't yet installed munt myself.

> But libao now works!

I wonder if it works right... Could you please check the
presence of this bug:
https://bugzilla.redhat.com/attachment.cgi?id=992924
in the sources?

> Hmm, if I SIGSTOP dosemu and then SIGCONT, libao gets out of sync... After printing this:

This is not a libao-related problem.
dosemu is set up to follow the real time, so if you
pause it, there is a timing gap that it tries to close by
speeding things up. It is needed so that the DOS time to
always stay correct. You can't easily move the DOS
time; you need to generate the needed amount of timer
interrupts too.
So while some surgery to add the gap-less timing source
is possible, my suggestion is simply to not SIGSTOP dosemu.

> [00000000] @ FFFF:0000 no PCM effect processors initialized

You may want to install ladspa to "fix" this.

**Comment by runderwo** [comment]
@stsp If I left wolf3d running, eventually it blew up with:
```
dosemu.bin: sndpcm.c:916: calc_idxs: Assertion `pcm.stream[i].buf_cnt >= pl->last_cnt[i]' failed.
```

If all the missed timer interrupts are fired to catch up, why does the sound go out of sync still?

**Comment by runderwo** [comment]
@stsp, none of the battr related code in ao_plugin_open() exists in ubuntu's libao-1.1.0.

**Comment by jingber** [comment]
@runderwo - I had a KVM switch die, so I have to setup the DOSEmu/MUNT stack on another machine.

When I played briefly with your mpu401 branch (and I will verify again this is what I was actually running), I'll provide the debug logs.  When I tried this branch, anything using an MPU-401 (standard mode or UART) would lock up immediately when the first note was expected.

If anyone prefers, I can move these reports regarding your branch to the appropriate github page.  I don't mind either way.

**Comment by stsp** [comment]
@runderwo Please fill the separate report about this
assert fail, and there we will talk about libao/timers/etc.

The interesting thing that may be related to this thread,
is that I can't get sysexes to pass neither to fluidsynth,
nor even to a midi file. Seems sysexes were never
tested before.

**Comment by runderwo** [comment]
@stsp alsa is passing the sysex differently (different SEQ command type) in uart vs intelligent mode.  I have traced it but not finished debugging.

@jingber please set $_mpu_irq=(2) and try again

**Comment by stsp** [comment]
I've got sysexes passed to fluidsynth with this patch:
http://sourceforge.net/p/fluidsynth/tickets/150/
Still no sysexes are passed to midi file, so I can't
make a capture of the problematic events.

**Comment by jingber** [comment]
[QUOTE]@jingber please set $_mpu_irq=(2) and try again with my branch (I noticed yesterday that I forgot to set it as default)[/QUOTE]

**Comment by runderwo** [comment]
Ok, the intelligent mode driver I copied from the earlier Silpheed release programs the MT32 partials one at a time, but the PS/1 version's UART driver concatenates all the partial programs into large sysexes.  In one of these cases it passes a sysex of 264 bytes, which is greater than the maximum 256 bytes of an event, so alsa splits it up into multiple events.
```        err = snd_midi_event_new(256, &virt->midi_event);```

I hoped to quickly fix this in munt by handling the split sysex event, but the problem is that alsa throws away the continued sysex event in snd_midi_event_decode() because it doesn't contain a valid start of a MIDI command (since it starts in the middle of a previous sysex payload).

I don't see any alternative to make this particular PS/1 Silpheed MT-32 driver work over the ALSA interface without convincing the developers that the hardcoded event buffer size is too small and should be configurable through the midi params, and waiting for new releases.

**Comment by runderwo** [comment]
There's also the "ALSA & OSS4 raw MIDI ports" section of munt's README.txt.  Using that would skip the virtual interface with the hardcoded event buffer size, but has other drawbacks; also, I have no idea how to use it right now.

**Comment by stsp** [comment]
> I hoped to quickly fix this in munt by handling the split sysex event, but the problem is that alsa throws
> away the continued sysex event in snd_midi_event_decode() because it doesn't contain a valid start
> of a MIDI command 

I haven't installed munt myself, but reading the alsa code:

			if (virt->in_event->type == SND_SEQ_EVENT_SYSEX) {
				virt->in_buf_ptr = virt->in_event->data.ext.ptr;
				virt->in_buf_size = virt->in_event->data.ext.len;
			} else {
				virt->in_buf_ptr = virt->in_tmp_buf;
				virt->in_buf_size = snd_midi_event_decode(virt->midi_event,
									  virt->in_tmp_buf,
									  sizeof(virt->in_tmp_buf),
									  virt->in_event);
			}

It seems snd_midi_event_decode() should not be in a way,
because the other branch of the condition is taken. Maybe
virt->in_event->type is reset earlier than needed, but I looked
through snd_midi_event_encode_byte() and it seems to be
doing the right thing too. So I would still expect that munt doesn't
handle the split event. Or if there is an alsa problem, could you
please explain it a bit further?

**Comment by stsp** [comment]
After a lot of patching, I've finally captured the midi
file with all sysexes (hopefully!).
Can someone please play it with munt and see how it goes?
[a.mid.zip](https://github.com/stsp/dosemu2/files/65659/a.mid.zip)

**Comment by stsp** [comment]
No, sounds still worse than the one on video...

**Comment by runderwo** [comment]
@stsp I saw the split sysex event emitted on the wire, and the missing second seq event incoming to munt.  alsa is throwing it away somewhere.

**Comment by stsp** [comment]
> alsa is throwing it away somewhere.

On your branch or on my branch?
I suspect we are still working on a different branches,
which is quite sad.

**Comment by runderwo** [comment]
On my branch, rebased to your devel as of last night.
It has nothing to do with my changes or anything in dosemu.
ALSA on client side (dosemu) generates 2 events from one long sysex.  They go on the wire and are correct in their contents.  By the time munt's event loop consumes the event, there is only one event that does not contain the EOX.

Different branches is not sad.  It's better that I keep my changes away most of the time :)

**Comment by stsp** [comment]
Yes, the events are split, but I thought my
branch fixes the loss of one part by disabling
the nonblock mode. If that was not enough,
I'll probably have to replicate this munt setup
myself...

**Comment by runderwo** [comment]
Don't bother, it's a hassle to set everything up for debugging and to log at the proper points.  I will probably solve it tonight, hopefully as a workaround rather than a patch to alsa-lib.

**Comment by runderwo** [comment]
The problem was not in alsa, I was comparing wrong parts of the logs late last night.  The problem was in Munt and is fixed with this PR
https://github.com/munt/munt/pull/22

**Comment by stsp** [comment]
@runderwo Congrats, your patch seems to be
already applied. Thanks!
Would you mind to also fill in the alsa bugreport
for overall completeness?

**Comment by stsp** [comment]
OK, and I finally managed to record the silpheed midi
music with all sysexes. So for someone's collection, here it is:
[aa.mid.zip](https://github.com/stsp/dosemu2/files/66654/aa.mid.zip)
Presumably, the sysex handling is now fixed everywhere
except for alsa (bug report should be opened).

**Comment by runderwo** [comment]
I'm not sure it's really a bug in ALSA.  So as not to block other realtime users, a large sysex, which is always of unknown length, has to be packetized at some arbitrary boundary.  Maybe the bug is rather that the documentation doesn't mention that a sequencer client must handle this case?

**Comment by stsp** [comment]
The bug here is that it seems alsa sequencer
interface allows up to 4096 bytes (and dosbox works),
while the rawmidi allows only 256 - unacceptable
disparity.
Also I don't think sysexes are opaque for all
users. Consider yourself a munt: the sysex is
not opaque for you then, because you need to
parse it fully. Obviously you can't parse it by
parts - you need the whole thing. And also
obviously you know the largest possible sysex
you support. I think the largest possible sysex
for munt is around 300bytes (we can ask its
authors to make sure). So why would they ever
deal with the larger sysexes, if they do not support
them anyway? So it would be very convenient
for them to have the alsa buffer larger than the
biggest sysex they support. Given that all other
software (and even alsa itself in its sequencer API)
fulfil this need, I think rawmidi should be bugreported.
Anyway, its of course up to you to decide.

**Comment by runderwo** [comment]
I suppose that they take the care to split an oversize sysex into multiple events is what makes me think generating multiple sequencer events is by design.
I'll send an email to alsa-devel@ (since their bugtracker seems to have been removed and this isn't a kernel issue).

**Comment by stsp** [comment]
Splitting itself is _probably_ by design.
See, for example, this:
http://mailman.alsa-project.org/pipermail/alsa-devel/2015-March/088852.html
It seems, sequencer interface throws away big
sysexes (over 4096), rather than fragmenting them.
But in any case, even if the fragmentation is by
design, the artificially small buffer size if likely not.
It should either be configurable by an app, or defaulted
to the value that will never require fragmenting (like
the sequencer interface seems to do).

**Comment by runderwo** [comment]
Well, I asked them whether to document the fixed size for users of snd_seq_input_event(), or to make the event buffer size a param like the dev buffer size.  But my mail is stuck in mod queue.

**Comment by stsp** [comment]
Ryan, so could you please check if we have now
with my additions to devel and your fix to munt, an
out-of-the-box working silpheed?
If so - we should close this.

**Comment by runderwo** [comment]
Yes, it works, but I have to disable fluidsynth at compile time or all midi output goes to it regardless of $_midi_driver.  Also, dosemu creates two ALSA midi connections:
client 129: 'Client-129' [type=user]
    0 'Virtual RawMIDI '
client 130: 'Client-130' [type=user]
    0 'Virtual RawMIDI '

I was able to get the correct output by connecting the first one (129:0) to munt.

I have no idea what was intended, but maybe you do.

Also, blaster.com no longer actually sets BLASTER and MIDI env vars at all.

**Comment by stsp** [comment]
> Yes, it works, but I have to disable fluidsynth at compile time

OK, the programmers should not test things - users should. :)

> $_midi_driver

$_midi_synth

> Also, dosemu creates two ALSA midi connections:

You still have redirected "default", right?

> I have no idea what was intended, but maybe you do.

OK, some documentation will help.

> Edit: now I see /e was added, so it works after all.

So I forgot to document the new /S option. It is now fixed.
Does it help?

**Comment by runderwo** [comment]
Sorry this took so long.
If I set $_midi_synth = mt32 in the config, there's a port conflict:
SB: mpu401 irq set to 5
PCM: registering player: SB REC
PCM: init
PCM: registering player: Sound Output: libao
PCM: registering player: Sound Output: libao wav writer
PCM: registering recorder: Sounds input: alsa
no PCM effect processors initialized
Initializing SDL sound output
PCM: Initializing selected plugin: Sound Output: SDL device: OK
PCM: Initializing pass-through plugin: SB REC: OK
PCM: Stream 0 allocated for "PCM LINE IN"
PCM: Initializing plugin: Sounds input: alsa (w=0): OK
PCM: Stream 1 allocated for "SB DAC"
PCM: setting flag 1 for stream 1 (SB DAC)
PCM: Stream 2 allocated for "SB DMA"
PCM: setting flag 4 for stream 2 (SB DMA)
PCM: registering recorder: MIDI Output: FluidSynth device
MIDI Output: named pipe: unable to open /home/nemesis/.dosemu/run/dosemu-midi for writing (No such device or address), will continue trying
PCM: Initializing pass-through plugin: MIDI Output: named pipe: OK
PCM: Bypassing plugin: MIDI Output: TiMidity++ plugin: (2 < 3)
Your fluidsynth is too old
fluidsynth: warning: Failed to pin the sample data to RAM; swapping is possible.
fluidsynth: loaded soundfont /usr/share/sounds/sf2/FluidR3_GM.sf2 ID=1
PCM: Stream 3 allocated for "MIDI"
PCM: Initializing plugin: MIDI Output: FluidSynth device (w=3): OK
PCM: Initializing plugin: MIDI Output: ALSA virtual device (for MT32) (w=0): OK
MIDI Input: named pipe: unable to open /home/nemesis/.dosemu/run/dosemu-midi_in for reading: No such file or directory
PCM: Initializing plugin: MIDI Input: named pipe (w=0): Failed
SB: SB Initialisation
PORT: registered "SB Emulation" handle 0x1c [0x0220-0x0233] fd=-1
SB: Mixer init
SB: OPL3 Initialization
PORT: registered "OPL3" handle 0x1d [0x0388-0x038b] fd=-1
PCM: Stream 4 allocated for "Adlib"
MPU401: MPU-401 Initialisation
ERROR: PORT: IRQ 5 conflict.  IO devices SB Emulation & Midi Emulation
ERROR: PORT: IRQ 5 conflict.  IO devices SB Emulation & Midi Emulation
ERROR: MPU-401: Cannot registering port handler
ERROR: MPU-401: Cannot registering port handler
MPU401: MPU-401 Initialisation - Base 0x330
SB: Initialisation - Base 0x220
SB: Resetting SB DSP
MFS: couldn't find root path /media/CDROM
timer_interrupt_rate count 6629, requested 180 Hz, granted 182 Hz
timer_interrupt_rate count 65536, requested 18.2 Hz, granted 18.2 Hz
PORTb: 331 not available for read
PORTb: 331 not available for read
PORTb: 331 not available for read
[..]

**Comment by runderwo** [comment]
Removing the $_midi_synth option from config and setting blaster /s mt32 before running silpheed with the PS/1 version MT-32 driver, it works correctly with $_midi_driver = alsa_virtual at least.  So the only remaining thing to do is integrate libmt32 directly or connect to munt automatically to avoid a manual aconnect command.

**Comment by stsp** [comment]
> If I set $_midi_synth = mt32 in the config, there's a port conflict:

Thanks, should now be fixed.

> with $_midi_driver = 

Please stop using silly options, or I'll remove them again. :)

> So the only remaining thing to do is integrate libmt32 directly

Looks impossible so far:
http://sourceforge.net/p/munt/support-requests/4/

> or connect to munt automatically

By dosemu startup script?
It is also interesting if they could just read the midi stream
from stdin and output the pcm to stdout. I coded that for
timidity, and now this is the way dosemu uses timidity.

**Comment by runderwo** [comment]
> Please stop using silly options, or I'll remove them again. :)

Ok, I removed the option.  It works anyway (if I use aconnect).  Clearly I don't understand what the intent is.  The virtual port is always enabled now?

> By dosemu startup script?

No, by searching the connections by name via alsa API and then connecting the dosemu virtual port to it.

**Comment by stsp** [comment]
> Clearly I don't understand what the intent is. The virtual port is always enabled now?

In case you set synth mode to mt32.
This is a poor-man's support of course.
Maybe later we'll have a directly used plugin.

> No, by searching the connections by name via alsa API and then connecting the dosemu virtual port to it.

This means someone have already started munt properly, right?
Which IMHO means he could as well call aconnect.
Why such a complexity if the script can do both things?

**Comment by runderwo** [comment]
> Why such a complexity if the script can do both things?

I run munt as part of a desktop session, like timidity daemon.
dosemu comes and goes (especially because of unavoidable crashes).

**Comment by stsp** [comment]
OK, added another silly option.
Now you can do
$_snd_plugin_params = "alsa_virtual:dev_name=munt"
Does this work for you?

**Comment by stsp** [comment]
Note that as soon as munt can work as a lib,
dosemu will start the private instance of munt,
no matter what you do "as part of a desktop session".
So all you do is wrong, but as a temporary solution
I have to add silly options for you...

**Comment by runderwo** [comment]
I don't need you to tell me that all I do is wrong.  Also, you don't have to do anything.
To clarify, this is what I was talking about.
https://github.com/stsp/dosemu2/pull/141

**Comment by stsp** [comment]
> I don't need you to tell me that all I do is wrong.

Last time I did, you've found that you didn't install
fluidsynth-dev and have old timidity. This time you
still mention timidity daemon, so it looks like you are
still forcing dosemu to respect your setup rather than
to just let it do the things the way it wants.

> Also, you don't have to do anything.

In this case you use your unmerged forks, so I'll better do.

> To clarify, this is what I was talking about.

I understand what you want to achieve, but I don't
understand what does it give us. Why can't you create
the virmidi devices so that dosemu is to just use name
"virtual" with rawmidi interface?

> one could more generically look for a port with SND_SEQ_PORT_TYPE_MIDI_MT32

This looks interesting.
Perhaps it can be implemented with the separate plugin
with this higher priority than the rawmidi one?

**Comment by stsp** [comment]
You can simply create the persistent rawmidi names
with .asoundrc. So I don't see the clear advantage in
what you propose. Of course searching for SND_SEQ_PORT_TYPE_MIDI_MT32
sounds also quite nice, so maybe we can have the both ways.

**Comment by runderwo** [comment]
>> it looks like you are
still forcing dosemu to respect your setup rather than
to just let it do the things the way it wants.

>> Why can't you create                                                                                                                                                           
   the virmidi devices so that dosemu is to just use name                                                                                                                                                          
   "virtual" with rawmidi interface? 

Because the program can work *without* manually hacking obscure configuration files in both instances (whether midi synth is private or not).  And what's the case for using non-virtual rawmidi anyway?

**Comment by stsp** [comment]
> Because the program can work without manually hacking
> obscure configuration files in both instances (whether midi synth is private or not).

But if, as you say, you set up it as a part of your
desktop environment, why can't you set up the
names in /etc/asound.conf too?
Or just provide an obscure names to dosemu, btw,
what would that be by default? Something like "virtual:0,1"?
I am not saying this is exceptionally convenient.
What I am saying is rather that this is a temporary
solution anyway, so I wouldn't invest too much time
into in (but SND_SEQ_PORT_TYPE_MIDI_MT32
may be worth a try).

> And what's the case for using non-virtual rawmidi anyway?

For MIDI-capable HW I guess. Either synth or just an uart.
You have a wave blaster, don't you? It will work.

**Comment by stsp** [comment]
> And what's the case for using non-virtual rawmidi anyway?

You are probably asking about mt32 only.
In this case I guess there is not much of a use for non-virtual
rawmidi. However SND_SEQ_PORT_TYPE_MIDI_GM may
also be possible. But then again, for GM I would suppose
"default" is more than enough...
I asked in munt tracker about the possibility of using it via
stdin/stdout pipes. If this is not too difficult, there will be no
need to hack the alsa plugin.

**Comment by stsp** [comment]
OK, the munt plugin is finally implemented.
I think this can be closed.

**Comment by stsp** [comment]
> @stsp If I left wolf3d running, eventually it blew up with:
> dosemu.bin: sndpcm.c:916: calc_idxs: Assertion `pcm.stream[i].buf_cnt >= pl->last_cnt[i]' failed.

@runderwo yes, this is a bug. I think you hit an integer
overflow: buf_cnt is a flat counter, never decrements. But
I've never seen that assert myself, perhaps you played Wolf
for quite too long. :) (or left it playing overnight)
I am not sure what would be the right fix here.
All players are periodically sampling this value, and if one
wants to implement some garbage-collect for it, he will need
to traverse all players in a thread-safe manner and decrement
the last-sampled value (finding the minimum sampled value
first, to not create an underflow). This is clearly sub-optimal.
Or maybe some completely different logic can be invented
that does not need a flat counter, but first the current one
have to be understood. :)
I think I'll just promote buf_cnt to "long long", which is supposed
to not overflow for a duration of a human life.
Otherwise, patches welcome.

**Comment by stsp** [comment]
Simple calculations show that at rate 44100 the
counter can overflow in about 13 hours.
Have you really played wolf for so long? :)
"long long" seems to survive for a half of million
years. I don't think someone will play wolf for half
a million years, so the fix should be good.

**Comment by stsp** [comment]
> Simple calculations show that at rate 44100 the
> counter can overflow in about 13 hours.

Or about 7 hours in stereo mode. Makes sense.
